contracts/pair_anchor/src/contract.rs

```diff
@@ -0,0 +1,860 @@
+            to_binary(&query_reverse_simulation(deps, ask_asset)?)
+        }
+        QueryMsg::CumulativePrices {} => to_binary(&query_cumulative_prices(deps, env)?),
+        QueryMsg::Config {} => to_binary(&query_config(deps)?),
+    }
+}
+
+/// ## Description
+/// Returns information about the pair contract in an object of type [`PairInfo`].
+/// ## Params
+/// * **deps** is an object of type [`Deps`].
+pub fn query_pair_info(deps: Deps) -> StdResult<PairInfo> {
+    let config: Config = CONFIG.load(deps.storage)?;
+    Ok(config.pair_info)
+}
+
+/// ## Description
+/// Returns the amounts of assets in the pair contract as well as the amount of LP
+/// tokens currently minted in an object of type [`PoolResponse`].
+/// ## Params
+/// * **deps** is an object of type [`Deps`].
+pub fn query_pool(deps: Deps) -> StdResult<PoolResponse> {
+    let config: Config = CONFIG.load(deps.storage)?;
+    let (assets, total_share) = pool_info(deps, config)?;
+
+    let resp = PoolResponse {
+        assets,
+        total_share,
+    };
+
+    Ok(resp)
+}
+
+/// ## Description
+/// Returns the amount of assets that could be withdrawn from the pool using a specific amount of LP tokens.
+/// The result is returned in a vector that contains objects of type [`Asset`].
+/// ## Params
+/// * **deps** is an object of type [`Deps`].
+///
+/// * **amount** is an object of type [`Uint128`]. This is the amount of LP tokens for which we calculate associated amounts of assets.
+pub fn query_share(_deps: Deps, _amount: Uint128) -> StdResult<Vec<Asset>> {
+    // let config: Config = CONFIG.load(deps.storage)?;
+    // let (pools, total_share) = pool_info(deps, config)?;
+    // let refund_assets = get_share_in_assets(&pools, amount, total_share);
+
+    Ok(vec![])
+}
+
+/// ## Description
+/// Returns information about a swap simulation in a [`SimulationResponse`] object.
+/// ## Params
+/// * **deps** is an object of type [`Deps`].
+///
+/// * **offer_asset** is an object of type [`Asset`]. This is the asset to swap as well as an amount of the said asset.
+pub fn query_simulation(deps: Deps, offer_asset: Asset) -> StdResult<SimulationResponse> {
+    let config: Config = CONFIG.load(deps.storage)?;        
+    let pools: [AssetInfo; 2] = config.pair_info.asset_infos;
+
+    if offer_asset.info.equal(&pools[0]) {
+    } else if offer_asset.info.equal(&pools[1]) {
+    } else {
+        return Err(StdError::generic_err(
+            "Given offer asset doesn't belong to pairs",
+        ));
+    }
+
+    let result: StateResponse = deps.querier.query_wasm_smart(
+        config.anchor_market_addr, 
+        &AnchorQueryMsg::State { block_height: None }
+    )?;
+
+    let return_amount;
+    let offer_amount: Uint256 = offer_asset.amount.into();
+
+    if offer_asset.is_native_token() {
+        return_amount = Uint128::from(offer_amount / result.prev_exchange_rate);
+    } else {
+        return_amount = Uint128::from(offer_amount * result.prev_exchange_rate);
```

Comment:
not sure if this should include some kind of tax for the calculation and how recent the prev_exchange_rate is from the anchor contract.

Reply:
Maybe it would be better to use
 terra1sepfj7s0aeg5967uxnfk4thzlerrsktkpelm5s/store?query_msg=eyJlcG9jaF9zdGF0ZSI6e319
{"epoch_state":{}}